src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LanguageServerProjectLoader.cs

```diff
@@ -273,4 +304,54 @@ async Task LogDiagnosticsAsync(ImmutableArray<DiagnosticLogItem> diagnosticLogIt
             await toastErrorReporter.ReportErrorAsync(worstLspMessageKind, message, cancellationToken);
         }
     }
+
+    /// <summary>
+    /// Creates a <see cref="LoadedProject"/> which has bare minimum information, but, which documents can be added to and obtained from.
+    /// </summary>
+    protected LoadedProjectSet AddLoadedProjectSet(string projectPath)
+    {
+        var projectSet = new LoadedProjectSet([], new SemaphoreSlim(1), new CancellationTokenSource());
+        return _loadedProjects.GetOrAdd(projectPath, projectSet);
+    }
+
+    protected async ValueTask TryUnloadProjectSetAsync(string projectPath)
+    {
+        if (_loadedProjects.TryRemove(projectPath, out var loadedProjectSet))
+        {
+            using var _ = await loadedProjectSet.Semaphore.DisposableWaitAsync();
```

Comment:
exposing the semaphore, and then needing the caller to manage it seems brittle and hard to reason about.  is tehre a reason a LoadedProject is not just a normal sealed calss, with private state, and contracts, that this type can call into ?

Reply:
It's guarding the array itself, as you don't want somebody else creating new projects and adding to it.